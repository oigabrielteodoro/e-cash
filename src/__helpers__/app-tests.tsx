import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
