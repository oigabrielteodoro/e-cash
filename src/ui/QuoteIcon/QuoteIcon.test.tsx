import React from 'react'

import { render } from '__helpers__/app-tests'

import { QuoteIcon } from '.'

describe('QuoteIcon', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<QuoteIcon />)

    expect(container).toMatchSnapshot()
  })
})
