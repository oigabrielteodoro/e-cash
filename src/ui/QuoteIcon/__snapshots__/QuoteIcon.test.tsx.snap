// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteIcon should be able match snapshot 1`] = `
.c0 {
  position: relative;
  margin-bottom: 1.6rem;
  max-width: 5.2rem;
}

.c0 svg {
  -webkit-animation: lbWRkT 3s;
  animation: lbWRkT 3s;
}

.c1 {
  position: absolute;
  background: #cfd7e3;
  width: 0.2rem;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
  top: calc(100% + 1.6rem);
}

.c2 {
  height: 0.2rem;
  position: absolute;
  background: #cfd7e3;
  width: 4rem;
  top: calc(100% + 26rem);
  left: calc(100% + -2.5rem);
}

<div>
  <div
    class="c0"
  >
    <svg
      fill="none"
      height="52"
      viewBox="0 0 52 52"
      width="52"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="26"
        cy="26"
        fill="#3a67e1"
        r="26"
      />
      <path
        d="M25.2905 27.6775V31.9355C25.2905 32.5269 25.0836 33.0296 24.6696 33.4436C24.2556 33.8576 23.7529 34.0646 23.1615 34.0646H18.9034C18.312 34.0646 17.8094 33.8576 17.3954 33.4436C16.9814 33.0296 16.7744 32.5269 16.7744 31.9355V24.1291C16.7744 23.3603 16.9241 22.6266 17.2235 21.928C17.5229 21.2294 17.9276 20.625 18.4377 20.115C18.9478 19.6049 19.5521 19.2001 20.2507 18.9008C20.9493 18.6014 21.683 18.4517 22.4518 18.4517H23.1615C23.3537 18.4517 23.52 18.5219 23.6605 18.6623C23.801 18.8028 23.8712 18.9691 23.8712 19.1613V20.5807C23.8712 20.7729 23.801 20.9392 23.6605 21.0797C23.52 21.2201 23.3537 21.2904 23.1615 21.2904H22.4518C21.6682 21.2904 20.9992 21.5676 20.4448 22.122C19.8903 22.6765 19.6131 23.3455 19.6131 24.1291V24.4839C19.6131 24.7796 19.7166 25.031 19.9236 25.238C20.1306 25.4449 20.3819 25.5484 20.6776 25.5484H23.1615C23.7529 25.5484 24.2556 25.7554 24.6696 26.1694C25.0836 26.5834 25.2905 27.0861 25.2905 27.6775ZM35.226 27.6775V31.9355C35.226 32.5269 35.019 33.0296 34.6051 33.4436C34.1911 33.8576 33.6884 34.0646 33.097 34.0646H28.8389C28.2475 34.0646 27.7448 33.8576 27.3309 33.4436C26.9169 33.0296 26.7099 32.5269 26.7099 31.9355V24.1291C26.7099 23.3603 26.8596 22.6266 27.159 21.928C27.4584 21.2294 27.8631 20.625 28.3732 20.115C28.8833 19.6049 29.4876 19.2001 30.1862 18.9008C30.8848 18.6014 31.6185 18.4517 32.3873 18.4517H33.097C33.2892 18.4517 33.4555 18.5219 33.596 18.6623C33.7364 18.8028 33.8067 18.9691 33.8067 19.1613V20.5807C33.8067 20.7729 33.7364 20.9392 33.596 21.0797C33.4555 21.2201 33.2892 21.2904 33.097 21.2904H32.3873C31.6037 21.2904 30.9347 21.5676 30.3803 22.122C29.8258 22.6765 29.5486 23.3455 29.5486 24.1291V24.4839C29.5486 24.7796 29.6521 25.031 29.8591 25.238C30.0661 25.4449 30.3174 25.5484 30.6131 25.5484H33.097C33.6884 25.5484 34.1911 25.7554 34.6051 26.1694C35.019 26.5834 35.226 27.0861 35.226 27.6775Z"
        fill="white"
      />
    </svg>
    <div
      class="c1"
      style="height: 0px;"
    />
    <div
      class="c2"
      style="width: 0px;"
    />
  </div>
</div>
`;
